src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/ThreadScheduler.kt

```diff
@@ -0,0 +1,392 @@
+     */
+    fun blockThread(threadId: ThreadId, reason: BlockingReason) {
+        threads[threadId]!!.apply {
+            blockingReason = reason
+            state = ThreadState.BLOCKED
+        }
+    }
+
+    /**
+     * Unblocks the thread.
+     * The thread state becomes [ThreadState.ENABLED].
+     *
+     * @param threadId The identifier of the thread to unblock.
+     */
+    fun unblockThread(threadId: ThreadId) {
+        threads[threadId]!!.apply {
+            state = ThreadState.ENABLED
+            blockingReason = null
+        }
+    }
+
+    /**
+     * Aborts the thread.
+     * The thread state becomes [ThreadState.ABORTED].
+     *
+     * @param threadId The identifier of the thread to abort.
+     */
+    fun abortThread(threadId: ThreadId) {
+        threads[threadId]!!.apply {
+            state = ThreadState.ABORTED
+        }
+    }
+
+    /**
+     * Aborts all threads controlled by the scheduler.
+     *
+     * @see abortThread
+     */
+    fun abortAllThreads() {
+        for (thread in threads.values) {
+            if (thread.state == ThreadState.FINISHED)
+                continue
+            thread.state = ThreadState.ABORTED
+        }
+    }
+
+    /**
+     * Aborts all threads controlled by the scheduler, except the current thread.
+     *
+     * @see abortThread
+     */
+    fun abortOtherThreads() {
+        val currentThreadId = getCurrentThreadId()
+        for (thread in threads.values) {
+            if (thread.state == ThreadState.FINISHED || thread.id == currentThreadId)
+                continue
+            thread.state = ThreadState.ABORTED
+        }
+    }
+
+    /**
+     * Marks the specified thread as finished.
+     * The thread states transitions to [ThreadState.FINISHED].
+     *
+     * Generally should be called just before the thread is actually going to finish its execution.
+     *
+     * @param threadId The identifier of the thread to be marked as finished.
+     */
+    fun finishThread(threadId: ThreadId) {
+        threads[threadId]!!.apply {
+            state = ThreadState.FINISHED
+        }
+    }
+
+    /**
+     * Awaits the completion of a thread up to a specified timeout.
+     *
+     * @param threadId The identifier of the thread to await.
+     * @param timeoutNano The maximum time to wait in nanoseconds.
+     * @return The elapsed time in nanoseconds if the thread finishes or aborts;
+     *   -1 if the timeout is reached.
+     */
+    fun awaitThreadFinish(threadId: ThreadId, timeoutNano: Long): Long {
+        val threadData = threads[threadId]!!
+        // special handling of Lincheck test threads
+        if (threadData.descriptor.thread is TestThread) {
+            val elapsedTime = threadData.spinner.spinWaitTimedUntil(timeoutNano) {
+                threadData.state == ThreadState.FINISHED ||
+                // TODO: due to limitations of current implementation,
+                //   Lincheck test threads sometime end up in ABORTED state,
```

Comment:
If I remember correctly, `onThreadFinish` is not guaranteed to be called at the end of the thread execution. 
This can happen in case of timeout, or in case if some exception was thrown from the code and not properly handled. 

To fix this, we will need to modify `TestThreadExecutionGenerator` class, to modify the generated bytecode. Wrap the code in try-catch, and put `onThreadFinish` into `finally` block to guarantee it is always called at the end.